### DISCUSSION TITLE
Regarding error while computing the SemanticSegmentationTask with num_classes=2 at loss_ fun

### DISCUSSION THREAD

**Discussion by kethan52** [torchgeo/torchgeo#597] [General]
model =SemanticSegmentationTask(
    segmentation_model="unet",
    encoder_name="resnet18",
    encoder_weights=None,
    in_channels=10,
    num_classes=2,
    num_filters=64,
    loss="jaccard",
    ignore_zeros=False,
    learning_rate=0.1,
    learning_rate_schedule_patience=5,
)

trainer = pl.Trainer(

                  gpus=1,
                  max_epochs=10,
                  precision=16,
                  log_every_n_steps=1,
                  max_steps=2,
               )

trainer.fit(model,dl)

_____________________________________________________________________________________________________________________________________________
Using 16bit native Automatic Mixed Precision (AMP)
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/configuration_validator.py:133: UserWarning: You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.
  rank_zero_warn("You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.")
Missing logger folder: /content/lightning_logs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name          | Type             | Params
---------------------------------------------------
0 | model         | Unet             | 14.4 M
1 | loss          | JaccardLoss      | 0     
2 | train_metrics | MetricCollection | 0     
3 | val_metrics   | MetricCollection | 0     
4 | test_metrics  | MetricCollection | 0     
---------------------------------------------------
14.4 M    Trainable params
0         Non-trainable params
14.4 M    Total params
28.701    Total estimated model params size (MB)
Epoch 0: 0%
0/4 [00:00<?, ?it/s]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-15-c68a00784715>](https://localhost:8080/#) in <module>()
     21                )
     22 
---> 23 trainer.fit(model,dl)

36 frames
[/usr/local/lib/python3.7/dist-packages/torchmetrics/utilities/checks.py](https://localhost:8080/#) in _check_shape_and_type_consistency(preds, target)
    113     else:
    114         raise ValueError(
--> 115             "Either `preds` and `target` both should have the (same) shape (N, ...), or `target` should be (N, ...)"
    116             " and `preds` should be (N, C, ...)."
    117         )

ValueError: Either `preds` and `target` both should have the (same) shape (N, ...), or `target` should be (N, ...) and `preds` should be (N, C, ...).

**Comment by kethan52** [comment]
Like as it was the binary segmentation i tried cross-entropy even for that I am getting error can anyone help to solve this problem

**Comment by calebrob6** [comment]
What dataloader/datamodule are you using? Can you post the code / initialization?

**Comment by kethan52** [comment]
class CustomDataset(RasterDataset):
  filename_glob = "*ryone.tif"

  def __getitem__(self, query):
    a_max=0.1
    data=super().__getitem__(query)
    data["mask"]=transform(data)["image"][None,8,:,:]
    
    data["mask"] = (data["mask"]<a_max)*1
    data["mask"].float()
    
    del data["bbox"]
    return data
  

ds = CustomDataset(data_root)
sampler = RandomGeoSampler(ds, size=76.8, length=4,roi=None)
dl = DataLoader(ds, sampler=sampler, collate_fn=stack_samples)
print(dl)
print(sampler)
print(ds)

i used the 76.8 becouse each is dividing by 0.3 so i get the size as torch.Size([1, 1, 256, 256])

**Comment by calebrob6** [comment]
Can you print the shape of `batch["mask"]`, it should be (batch size, height, width) I think. E.g.:
```
for batch in dl:
    print(batch["mask"].shape)
    break
```

**Comment by kethan52** [comment]
yes, the shape of the mask is torch.Size([1, 1, 256, 256])

for x in dl:
    data = x['image']
    target =x['mask']
    print(data.shape)
    print(target.shape)
---------------------------------------------------------------------------------------------------
torch.Size([1, 10, 256, 256])
torch.Size([1, 1, 256, 256])
torch.Size([1, 10, 256, 256])
torch.Size([1, 1, 256, 256])
torch.Size([1, 10, 256, 256])
torch.Size([1, 1, 256, 256])
torch.Size([1, 10, 256, 256])
torch.Size([1, 1, 256, 256])

**Comment by calebrob6** [comment]
I think you need to change your dataloader to return masks of size `(batch size, height, width)`

**Comment by kethan52** [comment]
Thank you soo much @calebrob6 i think it worked well

**Comment by calebrob6** [comment]
You're welcome!

**Comment by kethan52** [comment]
But can I know why my epochs was not moving?

![Screenshot (1427)](https://user-images.githubusercontent.com/41123619/164943096-25ecec87-083f-49e6-be0c-8b8440fbf1a9.png)
